testing/src/test_node.rs

```diff
@@ -0,0 +1,310 @@
+// Copyright (C) 2019-2021 Aleo Systems Inc.
+// This file is part of the snarkOS library.
+
+// The snarkOS library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The snarkOS library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the snarkOS library. If not, see <https://www.gnu.org/licenses/>.
+
+use pea2pea::{
+    protocols::{Handshake, Reading, Writing},
+    Config,
+    Connection,
+    Node as Pea2PeaNode,
+    Pea2Pea,
+};
+use snarkos_external_types::{BlockLocators, Client, Environment, Message};
+use snarkvm::{dpc::testnet2::Testnet2, traits::Network};
+use tokio::{
+    io::{AsyncReadExt, AsyncWriteExt},
+    sync::Mutex,
+    task,
+};
+use tracing::*;
+
+use std::{
+    collections::HashSet,
+    convert::TryInto,
+    io,
+    net::{IpAddr, Ipv4Addr, SocketAddr},
+    sync::Arc,
+    time::Duration,
+};
+
+// Consts & aliases.
+const CHALLENGE_HEIGHT: u32 = 0;
+const PING_INTERVAL_SECS: u64 = 5;
+const PEER_INTERVAL_SECS: u64 = 3;
+const DESIRED_CONNECTIONS: usize = 10;
+const MESSAGE_VERSION: u32 = <Client<Testnet2>>::MESSAGE_VERSION;
+
+type SnarkosMessage = Message<Testnet2, Client<Testnet2>>;
+
+/// The test node; it consists of a `Node` that handles networking and `State`
+/// that can be extended freely based on test requirements.
+#[derive(Clone)]
+pub struct TestNode {
+    node: Pea2PeaNode,
+    state: State,
+}
+
+/// snarkOS state required for test purposes.
+#[derive(Clone, Default)]
+struct State {
+    /// The list of known peers; `Pea2Pea` already has its own internal peer handling, but
+    /// snarkOS recognizes nodes by their listening port; this collection facilitates
+    /// the peering experience during tests and aligns it with snarkOS logic.
+    // FIXME: use the `Disconnect` protocol to remove disconnected peers
+    peers: Arc<Mutex<HashSet<SocketAddr>>>,
```

Comment:
Does this mean pea2pea's internal state is ignored in this case?

Reply:
no, it isn't - `pea2pea` just doesn't forbid "two-sided" connections, which are not needed here; all the state in this object is on top of the internal networking state of `pea2pea` (which will still forbid duplicate connections, obvious self-connections etc.)